koku/api/report/queries.py

```diff
@@ -0,0 +1,331 @@
+
+        Returns:
+            (Integer): time relative value providing query scope
+
+        """
+        if self.time_scope_value:
+            return self.time_scope_value
+
+        time_scope_units = self.get_filter_data('time_scope_units')
+        time_scope_value = self.get_filter_data('time_scope_value')
+        if not time_scope_value:
+            if not time_scope_units:
+                time_scope_value = -10
+            elif time_scope_units == 'month':
+                time_scope_value = -1
+            else:
+                time_scope_value = -10
+        self.time_scope_value = int(time_scope_value)
+        return time_scope_value
+
+    def _get_time_frame_filter(self):
+        """Obtain time frame filter.
+
+        Returns:
+            (DateTime): start datetime for query filter
+            (DateTime): end datetime for query filter
+            (List[DateTime]): List of all interval slices by resolution
+
+        """
+        time_scope_value = self.get_time_scope_value()
+        time_scope_units = self.get_time_scope_units()
+        resolution = self.get_resolution()
+        this_hour = timezone.now().replace(microsecond=0, second=0, minute=0)
+        start = None
+        end = None
+        interval = []
+        if time_scope_units == 'month':
+            if time_scope_value == -1:
+                # get current month
+                start = this_hour.replace(microsecond=0, second=0, minute=0, hour=0, day=1)
+                end = ReportQueryHandler.next_month(start)
+            else:
+                # get previous month
+                end = this_hour.replace(microsecond=0, second=0, minute=0, hour=0, day=1)
+                start = ReportQueryHandler.previous_month(end)
+        else:
+            if time_scope_value == -10:
+                # get last 10 days
+                end = this_hour
+                start = ReportQueryHandler.n_days_ago(this_hour, 10)
+            else:
+                # get last 30 days
+                end = this_hour
+                start = ReportQueryHandler.n_days_ago(this_hour, 30)
+
+        if start and end and resolution == 'daily':
+            interval = ReportQueryHandler.list_days(start, end)
+        if start and end and resolution == 'monthly':
+            interval = ReportQueryHandler.list_months(start, end)
+        return (start, end, interval)
+
+    def _add_interval_data(self, interval, query_data):
+        """Add the interval data if not present in cost usage data.
+
+        Args:
+            interval    (List[DateTime]): Valid interval list sliced by resolution
+            query_data  (List(Dict)): Queried data
+        Returns:
+            (List[Dict]): List of data for all intervals sliced by resolution
+
+        """
+        current = {}
+        data = []
+        resolution = self.get_resolution()
+        if query_data:
+            current = query_data.pop()
+        while interval:
```

Comment:
A while loop is fine, but since a list is an iterable in Python, you will generally see a for loop instead `for interval in intervals`. That way you don't have to pop your intervals. Also if that interval is to be reused by another call to the function, it's been destroyed by pops.

```
>>> listy = [1,2,3]
>>> def go_pop(l):
...    while l:
...      l.pop()
...
>>> go_pop(listy)
>>> listy
[]
```

Likewise with the query data. If you don't want to alter the passed in list you could use a for loop with the builtin `enumerate`

```
for i, interval in enumerate(intervals):
    current = query_data[i]
```

Reply:
I can swap to using a for loop for interval. Was using pop mostly because these lists were potentially of length.
```
interval = [<date=1>, <date=2>, <date=3>, <date=4>]
query_data = [<date=2>]
```
Destruction of the lists didn't matter here.

Reply:
Nah, makes sense.